### PR TITLE
[Kernel] Adding bias epilogue support for `cutlass_scaled_mm`

### PR DESCRIPTION
**Pull request by ProExpertProg**
This PR adds support for a bias term using the new epilogue infrastructure from #5391. This is a part of the AQ kernel push (#3975) we're working on at Neural Magic. In future PRs, we will add support for asymmetric activation quantization by folding the activation zero-point into the bias.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: tests/kernels/test_cutlass.py
Comment (tlrmchlsmth) [inline tests/kernels/test_cutlass.py:104]: could you give it a reason? e.g. 
`pytest.skip(reason='bias is not yet supported by sm90 CUTLASS kernels')`

ditto for the other skips

File: CMakeLists.txt
Comment (tlrmchlsmth) [inline CMakeLists.txt:6]: Could you describe the difference between this and what was there before?

File: csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cu
Comment (tlrmchlsmth) [inline csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cu:139]: Does it make sense for `ScaledEpilogueAzp` to be a subclass of `ScaledEpilogue`?

I think if the types were more compositional than they are and if `ScaledEpilogue` remained encapulsated, it would. In this case though, `ScaledEpilogueAzp` really cracks open `ScaledEpilogue` and looks inside of it -- the fact that you're using `cutlass::multiply_add` to apply both the bias and one of the scales necessitates this I think.

In this case, I would suggest not having `ScaledEpilogueAzp` inherit from `ScaledEpilogue` and instead factoring out the ScaleA and ScaleB type definitions another way. Maybe we could add a more generic type called `EpilogueWithTensorExpansion` and define the types there?

**Review by tlrmchlsmth — COMMENTED** [review]
Overall, looks pretty good. I had one comment about the subclassing of `ScaledEpilogue`. 

One other point is that `azp` is not defined anywhere in the code and I think it will be confusing for readers -- maybe it can be renamed to the more familiar `bias`, and then in the linear layer we can make clear how we are folding a term to correct for the activations' zero point into the bias

File: CMakeLists.txt
Comment (ProExpertProg) [reply CMakeLists.txt:6]: Accidentally committed, required to build locally, happy to put in a separate PR. The reason is that `option` only supports booleans, this is the equivalent syntax for a string variable.

File: tests/kernels/test_cutlass.py
Comment (ProExpertProg) [reply tests/kernels/test_cutlass.py:104]: Will remove as we're going to add c3x in this pr anyway

File: csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cu
Comment (ProExpertProg) [reply csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cu:139]: I like that! I'll try it, and if too complex, I'll just repeat them

**Comment by cyang49** [comment]
@ProExpertProg I'll wait for this one to be finalized before refactoring #5390 and asking for reviews

**Comment by ProExpertProg** [comment]
@cyang49 we actually added c3x in this support so we can probably close yours unless you wanted to add any other features

**Review by tlrmchlsmth — COMMENTED** [review]
Could you change the name of this PR, since it adds bias epilogue support for CUTLASS 3.x as well now?

**Review by tlrmchlsmth — APPROVED** [review]

**Review by tlrmchlsmth — CHANGES_REQUESTED** [review]
Oh, I just noticed that `azp` is still everywhere in this PR. To reduce confusion, could you rename `bias_azp` to just bias everywhere?

**Comment by njhill** [comment]
> @cyang49 we actually added c3x in this support so we can probably close yours unless you wanted to add any other features

@ProExpertProg could you add @cyang49 as co-author on this one then :pray:

**Review by tlrmchlsmth — APPROVED** [review]

**Comment by robertgshaw2-neuralmagic** [comment]
@njhill @cyang49 would you guys mind providing a review?

File: csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cu
Comment (cyang49) [inline csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cu:262]: Is there a reason to remove the dtype checking of a_scales & b_scales? There are some that check this and some don't. Should this be moved to the entry function?

File: csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu
Comment (cyang49) [inline csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:109]: Why is this named `SUPER` when it's referring to the same struct type?

File: csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu
Comment (cyang49) [inline csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:182]: It'd be nice if there's some documentation about why the ArgumentType is structured this way. When I implement this, I only find examples in cutlass code and tried to go with this:
```
  args.epilogue.thread = {
      // ternary op: a_scales * (b_scales * (A*B)) + per-row bias
      a_args,  // a_scales
      {
          // binary op: b_scales * (A*B)
          b_args,  // b_scales
          {},      // acc
          {}       // binary args: multiplies
      },
      bias_args,  // bias
      {}          // ternary args: multiply_add
  };
```
I suppose there is some documentation about what can be omitted?

**Review by cyang49 — COMMENTED** [review]
adding some comments

File: csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cu
Comment (ProExpertProg) [reply csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cu:262]: Yeah, this became a templated function which does not know about `a_scales` and `b_scales` anymore. The new entry point now does the check ([below](https://github.com/vllm-project/vllm/pull/5560/files/97158de203a7769e29967c67cdf2ff190d4f695a#diff-adbdc77eb328148f4530aaeea7d938138824bff858a86cbbff2454443011dde2R347-R348))

File: csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu
Comment (ProExpertProg) [reply csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:109]: Nice catch

File: csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu
Comment (ProExpertProg) [reply csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:182]: They are just structs, like [this one](https://github.com/NVIDIA/cutlass/blob/637b15906358191cb4238af419d408a65819d7ec/include/cutlass/epilogue/fusion/sm90_visitor_load_tma_warpspecialized.hpp#L568-L571) (used for bias):
```
  struct Arguments {
    Element const* ptr_row = nullptr;
    Element null_default = Element(0);
    StrideMNL dRow = {};
  };
```

So they all have default arguments. For compute nodes, the arguments are a tuple of arguments of children.

I am inclined to keep it this way for conciseness as I don't think it's as complicated. But I'll defer to @tlrmchlsmth here.

**Comment by cyang49** [comment]
Thanks! LGTM

File: csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu
Comment (tlrmchlsmth) [reply csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:182]: I think I actually did a slightly better job of breaking it down in the c2x kernel than in the c3x kernel: https://github.com/vllm-project/vllm/blob/a7dcc62086ea751b46b4821c2811cf8ac83711bf/csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cu#L130-L133

Whitespace could be improved there, but that code is more explicit about what the types are, and I've found that makes it easier to extend and debug as compared to the examples in the CUTLASS repo

**Comment by robertgshaw2-neuralmagic** [comment]
@cyang49 can you approve?

**Review by cyang49 — APPROVED** [review]

File: csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu
Comment (cyang49) [inline csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:543]: Is there a reason to limit bias dtype to float32?
We have use cases where bias is fp16 or bf16.

**Comment by cyang49** [comment]
I'm testing this PR on H100. I'm getting Error Internal.
```
ok bias0
Traceback (most recent call last):
  File "/home/ccyang/github.ibm.com/bench_flash_attn/scripts/simple_test_cutlass_scaled_mm.py", line 19, in <module>
    y = cutlass_scaled_mm(a, bt, out_dtype=out_dtype, scale_a=scale_a, scale_b=scale_b, bias=bias)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ccyang/github.com/vllm/vllm/_custom_ops.py", line 34, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/ccyang/github.com/vllm/vllm/_custom_ops.py", line 236, in cutlass_scaled_mm
    torch.ops._C.cutlass_scaled_mm(out, a, b, scale_a, scale_b, bias)
  File "/net/storage149/mnt/md0/ccyang/miniforge3/envs/vllm/lib/python3.11/site-packages/torch/_ops.py", line 854, in __call__
    return self_._op(*args, **(kwargs or {}))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Error Internal
```
@ProExpertProg could you see if you can reproduce this?
```
import torch
import time
import vllm
from vllm._custom_ops import cutlass_scaled_mm
m = 4
n = 6400
k = 6144
dtype = torch.float8_e4m3fn
out_dtype = torch.float16

a = torch.empty(m, k).normal_(mean=0.0, std=0.5).to(dtype=dtype, device='cuda')
bt = torch.empty(n, k).normal_(mean=0.0, std=0.5).to(dtype=dtype, device='cuda').t()
scale_a = torch.ones((1,)).to(dtype=torch.float32, device='cuda')
scale_b = torch.ones((1,)).to(dtype=torch.float32, device='cuda')
bias = torch.empty((n,)).normal_(mean=0.0, std=0.5).to(dtype=torch.float32, device='cuda')

y = cutlass_scaled_mm(a, bt, out_dtype=out_dtype, scale_a=scale_a, scale_b=scale_b)
print("ok bias0")
y = cutlass_scaled_mm(a, bt, out_dtype=out_dtype, scale_a=scale_a, scale_b=scale_b, bias=bias)
print("ok bias1")
```

File: csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu
Comment (ProExpertProg) [reply csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:543]: No reason except pure implementation simplicity. Do you want to open a new PR that adds support for `fp16` and `bf16` biases?

**Comment by ProExpertProg** [comment]
@cyang49 just ran on an H100 system, got this output:
```
$ python test.py
ok bias0
ok bias1
```

We recently found an issue with shared memory footprint that resulted in a `RuntimeError: Error Internal`. It was resolved for `sm80` in #5744, but maybe you're running into the same issue on an H100? Are you running on an H100_80GB?

**Comment by cyang49** [comment]
> @cyang49 just ran on an H100 system, got this output:
> 
> ```
> $ python test.py
> ok bias0
> ok bias1
> ```
> 
> We recently found an issue with shared memory footprint that resulted in a `RuntimeError: Error Internal`. It was resolved for `sm80` in #5744, but maybe you're running into the same issue on an H100? Are you running on an H100_80GB?

I'll try to merge that and rebuild on my side.

Yes, I'm using H100 80GB.

**Comment by ProExpertProg** [comment]
I don't think merging #5744 will help as it only fixes sm80 code, and sm90 (Hopper) is used on the H100.

File: csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu
Comment (cyang49) [reply csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:543]: I can do that, maybe after this one is merged..?
My old PR #5390 has bias type templated depending on `ElementD`

**Comment by cyang49** [comment]
> I don't think merging #5744 will help as it only fixes sm80 code, and sm90 (Hopper) is used on the H100.

Can you try the script again with `CUDA_LAUNCH_BLOCKING=1`?
```
CUDA_LAUNCH_BLOCKING=1 python simple_test_cutlass_scaled_mm.py 
```

**Comment by ProExpertProg** [comment]
@cyang49 yeah that fails with the same error that you got. Investigating now

File: csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu
Comment (ProExpertProg) [reply csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:543]: @tlrmchlsmth is out today but maybe he can chime in if we want to tie the type of bias to the type of `ElementD`

Comment (tlrmchlsmth) [reply csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:543]: Missed this -- Let's change the bias type to be ElementD for all cases for this PR. This definitely what we want to do for the fp8 case. 

In a future PR, to handle the AZP correction term for the int8 case we have a couple of options:
* Fold it into the bias, using fp16/bf16 
* Upconvert the bias to fp32 and fold the AZP term into it
* Keep them separate, applying the AZP term inside of the bias.
We still need to experiment to see what the affect on accuracy from the first two is. So let's get this PR landed without waiting for those

**Comment by ProExpertProg** [comment]
@cyang49 thanks again for the bug report - I think during the refactor I missed something in the config. Could you re-approve the PR so we can get it merged?

**Review by cyang49 — APPROVED** [review]

**Review by robertgshaw2-redhat — APPROVED** [review]

**Pull request merged**
